hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/main/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpoint.java

```diff
@@ -17,40 +17,119 @@
+import org.apache.hadoop.hbase.replication.ReplicationPeer;
+import org.apache.hadoop.hbase.replication.ReplicationPeerConfig;
+import org.apache.hadoop.hbase.replication.ScopeWALEntryFilter;
+import org.apache.hadoop.hbase.replication.WALEntryFilter;
 import org.apache.hadoop.hbase.wal.WAL;
+import org.apache.hbase.thirdparty.com.google.common.collect.Lists;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.AbstractService;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-// TODO(remove BaseReplicationEndpoint extension).
 @InternalExtensionOnly
-public class HbaseToCloudBigtableReplicationEndpoint extends BaseReplicationEndpoint {
+public class HbaseToCloudBigtableReplicationEndpoint extends AbstractService
+    implements ReplicationEndpoint {
 
   private static final Logger LOG =
       LoggerFactory.getLogger(HbaseToCloudBigtableReplicationEndpoint.class);
 
   private final CloudBigtableReplicator cloudBigtableReplicator;
   private final HBaseMetricsExporter metricsExporter;
 
+  protected Context ctx;
+
   public HbaseToCloudBigtableReplicationEndpoint() {
     cloudBigtableReplicator = new CloudBigtableReplicator();
     metricsExporter = new HBaseMetricsExporter();
   }
 
+  @Override
+  public void init(Context context) throws IOException {
+    this.ctx = context;
+    if (this.ctx != null) {
+      ReplicationPeer peer = this.ctx.getReplicationPeer();
+      if (peer != null) {
+        peer.registerPeerConfigListener(this);
+      } else {
+        LOG.warn(
+            "Not tracking replication peer config changes for Peer Id "
+                + this.ctx.getPeerId()
+                + " because there's no such peer");
+      }
+    }
+  }
+
+  @Override
+  public boolean canReplicateToSameCluster() {
+    return false;
+  }
+
   @Override
   public UUID getPeerUUID() {
     return cloudBigtableReplicator.getPeerUUID();
   }
 
+  @Override
+  public WALEntryFilter getWALEntryfilter() {
+    ArrayList<WALEntryFilter> filters = Lists.newArrayList();
+    WALEntryFilter scopeFilter = getScopeWALEntryFilter();
+    if (scopeFilter != null) {
+      filters.add(scopeFilter);
+    }
+    WALEntryFilter tableCfFilter = getNamespaceTableCfWALEntryFilter();
+    if (tableCfFilter != null) {
+      filters.add(tableCfFilter);
+    }
+    if (ctx != null && ctx.getPeerConfig() != null) {
+      String filterNameCSV =
+          ctx.getPeerConfig().getConfiguration().get(REPLICATION_WALENTRYFILTER_CONFIG_KEY);
+      if (filterNameCSV != null && !filterNameCSV.isEmpty()) {
+        String[] filterNames = filterNameCSV.split(",");
+        for (String filterName : filterNames) {
+          try {
+            Class<?> clazz = Class.forName(filterName);
+            filters.add((WALEntryFilter) clazz.getDeclaredConstructor().newInstance());
+          } catch (Exception e) {
+            LOG.error("Unable to create WALEntryFilter " + filterName, e);
```

Comment:
I wonder if we should fail fast here? How do you communicate to the users that they are getting more entries than they expected since the filter is not recognized. This is especially true for custom filters that they may want to load.

Reply:
I suppose there are two ways:
1. Keep it as null (the default filter excludes the system tables) https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java#L330 here
2. Keep the implementation on par with HBaseReplicationEndpoint. 
I chose 2 here and keeping it sync with HBaseReplicationEndpoint

Reply:
ack, please add a comment listing this reason. Also, add tests for these filters.